### PR TITLE
r/aws_dynamodb_table_gsi

### LINKED ISSUE #671: No method to ignore changes in DynamoDB GSI
_This issue was originally opened by @cypai as hashicorp/terraform#13393. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
$ terraform -v
Terraform v0.8.8

### Affected Resource(s)
- aws_dynamodb_table

### Terraform Configuration Files
```hcl
resource "aws_dynamodb_table" "FooBar" {
    name = "FooBar"
    read_capacity = 10
    write_capacity = 10
    hash_key = "foo-id"
    attribute {
        name = "foo-id"
        type = "S"
    }
    attribute {
        name = "bar-id"
        type = "N"
    }
    global_secondary_index {
        name = "bar-index"
        hash_key = "bar-id"
        read_capacity = 10
        write_capacity = 10
        projection_type = "ALL"
    }
    lifecycle {
        ignore_changes = ["read_capacity", "write_capacity"]
    }
}
```

### References
- hashicorp/terraform#5617 

We would like to ignore read/write capacity for the GSI as well, but this configuration only ignores the read/write capacity of the top-level table. We weren't able to find any usable workaround, since we can't interpolate variables in the lifecycle, and wildcards are not supported.

Technically we could hardcode the randomly generated number like ```ignore_changes = ["global_secondary_index.3291674533.read_capacity"]```, but that's obviously a terrible solution since it would change every time we make a change.

### LINKED ISSUE #17096: New Resource: DyanmoDB Global Secondary Index (aws_dynamodb_table_gsi)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I have a project making use of a single-table design where sparse-index GSIs need to be added dynamically as the project's services expand. The lack of this resource is currently my only roadblock for having this fully modularized in Terraform (that is to say, for each new service, instead of the service's Terraform config itself being fully decoupled, I still have to manually add a new GSI via the table attributes to my base project Terraform config before applying the service config separately).

### New or Affected Resource(s)

* aws_dynamodb_table_gsi

### Potential Terraform Configuration

```hcl
# adapted from the abandoned pull request below
resource "aws_dynamodb_table_gsi" "gsi" {
  table_name = aws_dynamodb_table.table.id
  name = "gsi"
  hash_key {
    name = "TestGsiHashKey"
    type = "S"
  }
  write_capacity = 1
  read_capacity = 1
  projection_type = "KEYS_ONLY"
}
```

### References

This pull request (kindly authored by @ben-bourdin451) was looking to implement this resource, but seems it got stymied by backwards compatibility tests and has since been abandoned. I've been trying to carve some time to take a look at it myself but unfortunately my Golang fu is not presently what it could be.

* https://github.com/hashicorp/terraform-provider-aws/pull/6688

### PR DESCRIPTION
**Pull request by danquack**
Co-Authored-By: @ben-bourdin451. updated to v2 sdks from #6688

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates or Closes #17096
Relates or Closes #671

Output from acceptance testing:

```
$ TF_ACC=1 go test ././internal/service/dynamodb/... -v -count 1 -parallel 3  -run=TestAccAWSDynamoDbTableGsi_ -timeout 180m
=== RUN   TestAccAWSDynamoDbTableGsi_basic
=== PAUSE TestAccAWSDynamoDbTableGsi_basic
=== RUN   TestAccAWSDynamoDbTableGsi_updateCapacity
=== PAUSE TestAccAWSDynamoDbTableGsi_updateCapacity
=== RUN   TestAccAWSDynamoDbTableGsi_updateNonAttributes
=== PAUSE TestAccAWSDynamoDbTableGsi_updateNonAttributes
=== RUN   TestAccAWSDynamoDbTableGsi_delete
=== PAUSE TestAccAWSDynamoDbTableGsi_delete
=== CONT  TestAccAWSDynamoDbTableGsi_basic
=== CONT  TestAccAWSDynamoDbTableGsi_updateNonAttributes
=== CONT  TestAccAWSDynamoDbTableGsi_updateCapacity
--- PASS: TestAccAWSDynamoDbTableGsi_basic (280.10s)
=== CONT  TestAccAWSDynamoDbTableGsi_delete
--- PASS: TestAccAWSDynamoDbTableGsi_updateCapacity (309.75s)
--- PASS: TestAccAWSDynamoDbTableGsi_updateNonAttributes (691.06s)
--- PASS: TestAccAWSDynamoDbTableGsi_delete (431.54s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   714.664s
```

Contributor Notes:
Open for feedback...
- similar to aws security groups ingress/egress + security group rules, separating GSIs has an issue where attributes can be defined in different places. To avoid computed value differences when reading, I used a CustomizeDiff, but it clears the whole attribute. Not sure if you can clear just the attribute in the set in question.
- Along the same lines, i added a feature switch to manage it externally. Is this an acceptable approach?

### CODE REVIEW

**Comment by danquack** [comment]
@ewbankkit no need to review now. This is a WIP PR and will need some love based on feedback from the GitHub issues associated

File: internal/service/dynamodb/table.go [hunk 1]
```diff
@@ -129,40 +129,51 @@ func ResourceTable() *schema.Resource {
...
 			},
+			"manage_index_as_own_resource": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
 			"global_secondary_index": {
 				Type:     schema.TypeSet,
 				Optional: true,
+				Computed: true,
```
Comment (danquack) [inline internal/service/dynamodb/table.go:140]: Open for discussion, should we set the `Deprecated` bit?

**Comment by danquack** [comment]
@ewbankkit wondering if I could bump this for your review?

**Comment by ewbankkit** [comment]
@danquack Thanks for the work here :tada: :clap:.

Our concern with attempting to decompose the DynamoDB Table resource into multiple smaller resources (as has been done for S3 Bucket recently) is that the underlying AWS API doesn't cleanly support it (unlike the S3 API which does have separate API actions for the sub-resources).
Most noticeably the `attributes` property is shared between the Table and any GSIs and attempting to manage this property through both an `aws_dynamodb_table` resource and an associated `aws_dynamodb_table_gsi` resource will lead to continual differences being reported by Terraform in both resources and hence to practitioner confusion.

Attempting to address this (as I see you have done) via a `CustomizeDiff` function on the `aws_dynamodb_table` is potentially brittle and a source of practitioner (and maintainer) confusion.

For example, running the GSI tests for `aws_dynamodb_table` now reports errors:

```console
% make testacc TESTS=TestAccDynamoDBTable_gsi PKG=dynamodb ACCTEST_PARALLELISM=1
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/dynamodb/... -v -count 1 -parallel 1 -run='TestAccDynamoDBTable_gsi'  -timeout 180m
=== RUN   TestAccDynamoDBTable_gsiUpdateCapacity
=== PAUSE TestAccDynamoDBTable_gsiUpdateCapacity
=== RUN   TestAccDynamoDBTable_gsiUpdateOtherAttributes
=== PAUSE TestAccDynamoDBTable_gsiUpdateOtherAttributes
=== RUN   TestAccDynamoDBTable_gsiUpdateNonKeyAttributes
=== PAUSE TestAccDynamoDBTable_gsiUpdateNonKeyAttributes
=== CONT  TestAccDynamoDBTable_gsiUpdateCapacity
    table_test.go:847: Step 1/3 error: Error running apply: exit status 1
        
        Error: error creating DynamoDB Table: InvalidParameter: 1 validation error(s) found.
        - missing required field, CreateTableInput.AttributeDefinitions.
        
        
          with aws_dynamodb_table.test,
          on terraform_plugin_test.tf line 6, in resource "aws_dynamodb_table" "test":
           6: resource "aws_dynamodb_table" "test" {
        
--- FAIL: TestAccDynamoDBTable_gsiUpdateCapacity (6.15s)
=== CONT  TestAccDynamoDBTable_gsiUpdateNonKeyAttributes
    table_test.go:1044: Step 1/3 error: Error running apply: exit status 1
        
        Error: error creating DynamoDB Table: InvalidParameter: 1 validation error(s) found.
        - missing required field, CreateTableInput.AttributeDefinitions.
        
        
          with aws_dynamodb_table.test,
          on terraform_plugin_test.tf line 6, in resource "aws_dynamodb_table" "test":
           6: resource "aws_dynamodb_table" "test" {
        
--- FAIL: TestAccDynamoDBTable_gsiUpdateNonKeyAttributes (5.11s)
=== CONT  TestAccDynamoDBTable_gsiUpdateOtherAttributes
    table_test.go:915: Step 1/3 error: Error running apply: exit status 1
        
        Error: error creating DynamoDB Table: InvalidParameter: 1 validation error(s) found.
        - missing required field, CreateTableInput.AttributeDefinitions.
        
        
          with aws_dynamodb_table.test,
          on terraform_plugin_test.tf line 6, in resource "aws_dynamodb_table" "test":
           6: resource "aws_dynamodb_table" "test" {
        
--- FAIL: TestAccDynamoDBTable_gsiUpdateOtherAttributes (4.90s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb	22.120s
FAIL
make: *** [testacc] Error 1
```

DynamoDB tables are mission-critical resources at many practitioners’ organizations and we prefer to err on the side of caution when making significant changes to workflows built around this resource.

An AWS support ticket has been opened to start the conversation around making changes to the DynamoDB API to enable table index operations to be performed separately from the table itself, but as you can imagine this will not be a quick change.

Any limitations to the current `CustomizeDiff` functionality may be mitigated by eventual migration of this resource to the [`terraform-plugin-framework`](https://github.com/hashicorp/terraform-plugin-framework).

**Pull request closed without merging**